src/Features/Core/Portable/CodeRefactorings/SyncNamespace/AbstractChangeNamespaceService.cs

```diff
@@ -102,14 +108,77 @@ public override async Task<bool> CanChangeNamespaceAsync(Document document, Synt
             return !applicableContainers.IsDefault;
         }
 
+        public override async Task<Solution?> TryChangeTopLevelNamespacesAsync(
+            Document document,
+            string targetNamespace,
+            CancellationToken cancellationToken)
+        {
+            var syntaxFacts = document.GetRequiredLanguageService<ISyntaxFactsService>();
+            var syntaxRoot = await document.GetRequiredSyntaxRootAsync(cancellationToken).ConfigureAwait(false);
+
+            // Don't descend into anything other than top level declarations from the root.
+            // ChangeNamespaceService only controls top level declarations right now.
+            // Don't use namespaces that already match the target namespace
+            var originalNamespaceDeclarations = await GetTopLevelNamespacesAsync(document, cancellationToken).ConfigureAwait(false);
+
+            if (originalNamespaceDeclarations.Length == 0)
+            {
+                return null;
+            }
+
+            var semanticModel = await document.GetRequiredSemanticModelAsync(cancellationToken).ConfigureAwait(false);
+            var originalNamespaceName = semanticModel.GetDeclaredSymbol(originalNamespaceDeclarations.First()).Name;
+            var solution = document.Project.Solution;
+
+            // Only loop as many top level namespace declarations as we originally had. 
+            // Change namespace doesn't change this number, so this helps limit us and
+            // rule out namespaces that didn't need to be changed
+            for (var i = 0; i < originalNamespaceDeclarations.Length; i++)
+            {
+                var namespaceName = semanticModel.GetDeclaredSymbol(originalNamespaceDeclarations[i]).Name;
+                if (namespaceName != originalNamespaceName)
+                {
+                    // Skip all namespaces that didn't match the original namespace name that 
+                    // we were syncing. 
+                    continue;
+                }
+                syntaxRoot = await document.GetRequiredSyntaxRootAsync(cancellationToken).ConfigureAwait(false);
+
+                // Since the original namespaces were retrieved before the document was modified
+                // get the current top level namespaces. Since we're only renaming namespaces, the 
+                // number and index of each is the same.
+                var namespaces = await GetTopLevelNamespacesAsync(document, cancellationToken).ConfigureAwait(false);
+                Debug.Assert(namespaces.Length == originalNamespaceDeclarations.Length);
+
+                var namespaceToRename = namespaces[i];
+                solution = await ChangeNamespaceAsync(document, namespaceToRename, targetNamespace, cancellationToken).ConfigureAwait(false);
+                document = solution.GetRequiredDocument(document.Id);
+            }
```

Comment:
do we have tests that validate this works?  i.e. 3 namespaces in a file, and the middle one is named correctly (and should be skippeD)?   just want to make sure we're testing the corners :)

Reply:
I'm always down to add more tests :) 

I'll add one now.